File: src/main/java/org/modelingvalue/nelumbo/Query.java
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// (C) Copyright 2018-2025 Modeling Value Group B.V. (http://modelingvalue.org)                                        ~
//                                                                                                                     ~
// Licensed under the GNU Lesser General Public License v3.0 (the 'License'). You may not use this file except in      ~
// compliance with the License. You may obtain a copy of the License at: https://choosealicense.com/licenses/lgpl-3.0  ~
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on ~
// an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the  ~
// specific language governing permissions and limitations under the License.                                          ~
//                                                                                                                     ~
// Maintainers:                                                                                                        ~
//     Wim Bast, Tom Brus                                                                                              ~
//                                                                                                                     ~
// Contributors:                                                                                                       ~
//     Victor Lap                                                                                                      ~
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

package org.modelingvalue.nelumbo;

import java.io.Serial;
import java.util.Optional;

import org.modelingvalue.collections.Entry;
import org.modelingvalue.collections.List;
import org.modelingvalue.collections.Map;
import org.modelingvalue.collections.Set;
import org.modelingvalue.nelumbo.patterns.Functor;
import org.modelingvalue.nelumbo.syntax.ParseException;
import org.modelingvalue.nelumbo.syntax.ParserResult;

public final class Query extends Node implements Evaluatable {
    @Serial
    private static final long serialVersionUID = -6751904607718047038L;

    private InferResult       inferResult;

    public Query(Functor functor, List<AstElement> elements, Object... args) {
        super(functor, elements, args(elements, args));
    }

    @SuppressWarnings("unchecked")
    private static Object[] args(List<AstElement> elements, Object[] args) {
        Predicate nodePred = (Predicate) args[0];
        Predicate predicate = nodePred.setVariables(Predicate.literals(nodePred.variables()));
        Optional<List<List<Object>>> expected = (Optional<List<List<Object>>>) args[1];
        if (expected.isEmpty()) {
            return new Object[]{predicate};
        }
        List<Object> facts = expected.get().get(0);
        List<Object> falsehoods = expected.get().get(1);
        Object[] array = new Object[5];
        array[0] = predicate;
        array[1] = bindings(facts.filter(List.class).asList());
        array[2] = !facts.contains("..");
        array[3] = bindings(falsehoods.filter(List.class).asList());
        array[4] = !falsehoods.contains("..");
        return array;
    }

    @SuppressWarnings("unchecked")
    private static Set<Map<Variable, Object>> bindings(List<?> listListList) {
        Set<Map<Variable, Object>> set = Set.of();
        for (List<List<Object>> listList : (List<List<List<Object>>>) listListList) {
            Map<Variable, Object> map = Map.of();
            for (List<Object> list : listList) {
                map = map.put(((Variable) list.get(0)).literal(), (Node) list.get(1));
            }
            set = set.add(map);
        }
        return set;
    }

    @SuppressWarnings({"rawtypes", "unchecked"})
    @Override
    public List<Object> args() {
        List<Object> args = List.of(predicate());
        if (hasExpected()) {
            List<Object> trueList = List.of();
            for (Map<Variable, Object> binding : facts()) {
                List<List<Object>> list = List.of();
                for (Entry<Variable, Object> e : binding) {
                    list = list.add(List.of(e.getKey(), e.getValue()));
                }
                trueList = trueList.add(list);
            }
            if (!completeFacts()) {
                trueList = trueList.add("..");
            }
            List<Object> falseList = List.of();
            for (Map<Variable, Object> binding : falsehoods()) {
                List<List<Object>> list = List.of();
                for (Entry<Variable, Object> e : binding) {
                    list = list.add(List.of(e.getKey(), e.getValue()));
                }
                falseList = falseList.add(list);
            }
            if (!completeFalsehoods()) {
                falseList = falseList.add("..");
            }
            args = args.add(Optional.of(List.of(trueList, falseList)));
        } else {
            args = args.add(Optional.empty());
        }
        return args;
    }

    private Query(Object[] array) {
        super(array);
    }

    @Override
    protected Query struct(Object[] array) {
        return new Query(array);
    }

    @Override
    public Query set(int i, Object... a) {
        return (Query) super.set(i, a);
    }

    public Predicate predicate() {
        return (Predicate) get(0);
    }

    @SuppressWarnings("unchecked")
    public Set<Map<Variable, Object>> facts() {
        return (Set<Map<Variable, Object>>) get(1);
    }

    public boolean completeFacts() {
        return (java.lang.Boolean) get(2);
    }

    @SuppressWarnings("unchecked")
    public Set<Map<Variable, Object>> falsehoods() {
        return (Set<Map<Variable, Object>>) get(3);
    }

    public boolean completeFalsehoods() {
        return (java.lang.Boolean) get(4);
    }

    public boolean hasExpected() {
        return length() > 1;
    }

    @Override
    public void evaluate(KnowledgeBase knowledgeBase, ParserResult result) throws ParseException {
        Predicate predicate = predicate();
        InferResult found;
        try {
            found = predicate.infer().predicate(predicate);
        } catch (InconsistencyException ie) {
            result.addException(new ParseException(ie.getMessage(), ie.rule()));
            return;
        }
        if (hasExpected()) {
            toString();
            Set<Map<Variable, Object>> trueBindings = facts();
            Set<Predicate> truePredicates = trueBindings.map(b -> predicate.setBinding(b)).asSet();
            boolean completeFacts = completeFacts();
            Set<Map<Variable, Object>> falseBindings = falsehoods();
            Set<Predicate> falsePredicates = falseBindings.map(b -> predicate.setBinding(b)).asSet();
            boolean completeFalsehoods = completeFalsehoods();
            InferResult expected = InferResult.of(predicate, truePredicates, completeFacts, falsePredicates, completeFalsehoods, Set.of());
            if (!found.equals(expected) && !found.toString().equals(expected.toString())) {
                List<AstElement> astElements = astElements();
                result.addException(new ParseException("Expected result " + expected + ", found " + found, //
                        astElements.sublist(2, astElements.size()).toArray(i -> new AstElement[i])));
            }
        }
        inferResult = found;
    }

    public InferResult inferResult() {
        return inferResult;
    }

}

File: web-ts/src/Query.ts
/**
 * Query - logic queries with expected results.
 * @JAVA_REF org.modelingvalue.nelumbo.Query
 */

import { List, Map, Set } from 'immutable';
import type { AstElement } from './AstElement';
import { Node } from './Node';
import { Variable } from './Variable';
import type { Functor } from './patterns/Functor';
import { Predicate } from './logic/Predicate';
import { InferResult } from './InferResult';
import { ParseException } from './syntax/ParseException';
import type { KnowledgeBase, ParseExceptionHandler } from './KnowledgeBase';
import type { Evaluatable } from './Evaluatable';
import { Optional, some, none } from './patterns/OptionalPattern';

/**
 * Query - a query against the knowledge base with optional expected results.
 */
export class Query extends Node implements Evaluatable {
  private _inferResult: InferResult | null = null;

  constructor(functor: Functor, elements: List<AstElement>, ...args: unknown[]) {
    super(functor, elements, ...Query.processArgs(args));
  }

  private static processArgs(args: unknown[]): unknown[] {
    const nodePred = Predicate.predicate(args[0] as Node);
    const predBinding = nodePred.getBinding();
    const predicate = predBinding !== null
      ? nodePred.setVariables(Predicate.literals(predBinding))
      : nodePred;
    const expected = args[1] as Optional<List<List<unknown>>>;

    if (expected.isEmpty()) {
      return [predicate];
    }

    const expectedList = expected.get()!;
    const facts = expectedList.get(0) as List<unknown>;
    const falsehoods = expectedList.get(1) as List<unknown>;

    return [
      predicate,
      Query.bindings(facts.filter((x): x is List<List<unknown>> => List.isList(x)).toList()),
      !facts.contains('..'),
      Query.bindings(falsehoods.filter((x): x is List<List<unknown>> => List.isList(x)).toList()),
      !falsehoods.contains('..')
    ];
  }

  private static bindings(listListList: List<List<List<unknown>>>): Set<Map<Variable, unknown>> {
    let set = Set<Map<Variable, unknown>>();
    for (const listList of listListList) {
      let map = Map<Variable, unknown>();
      for (const list of listList) {
        map = map.set((list.get(0) as Variable).literal(), list.get(1));
      }
      set = set.add(map);
    }
    return set;
  }

  protected static fromDataQuery(data: unknown[], declaration?: Node): Query {
    const query = Object.create(Query.prototype) as Query;
    (query as unknown as { _data: unknown[] })._data = data;
    (query as unknown as { _declaration: Node | undefined })._declaration = declaration ?? query;
    return query;
  }

  protected override struct(data: unknown[], declaration?: Node): Node {
    return Query.fromDataQuery(data, declaration ?? this.declaration());
  }

  override args(): List<unknown> {
    const args: unknown[] = [this.get(0)];
    if (this.hasExpected()) {
      let trueList = List<unknown>();
      for (const binding of this.expectedFacts()) {
        let list = List<List<unknown>>();
        for (const [key, value] of binding.entries()) {
          list = list.push(List([key, value]));
        }
        trueList = trueList.push(list);
      }
      if (!this.completeFacts()) {
        trueList = trueList.push('..');
      }
      let falseList = List<unknown>();
      for (const binding of this.expectedFalsehoods()) {
        let list = List<List<unknown>>();
        for (const [key, value] of binding.entries()) {
          list = list.push(List([key, value]));
        }
        falseList = falseList.push(list);
      }
      if (!this.completeFalsehoods()) {
        falseList = falseList.push('..');
      }
      args.push(some(List([trueList, falseList])));
    } else {
      args.push(none<List<List<unknown>>>());
    }
    return List(args);
  }

  override set(i: number, ...a: unknown[]): Node {
    return super.set(i, ...a);
  }

  /**
   * Get the predicate being queried.
   */
  predicate(): Predicate {
    return Predicate.predicate(this.get(0) as Node);
  }

  /**
   * Get expected fact bindings.
   */
  expectedFacts(): Set<Map<Variable, unknown>> {
    return this.get(1) as Set<Map<Variable, unknown>>;
  }

  /**
   * Check if expected facts are complete.
   */
  completeFacts(): boolean {
    return this.get(2) as boolean;
  }

  /**
   * Get expected falsehood bindings.
   */
  expectedFalsehoods(): Set<Map<Variable, unknown>> {
    return this.get(3) as Set<Map<Variable, unknown>>;
  }

  /**
   * Check if expected falsehoods are complete.
   */
  completeFalsehoods(): boolean {
    return this.get(4) as boolean;
  }

  /**
   * Check if this query has expected results.
   */
  hasExpected(): boolean {
    return this.length() > 1;
  }

  /**
   * Evaluate this query against a knowledge base.
   */
  evaluate(knowledgeBase: KnowledgeBase, handler: ParseExceptionHandler): void {
    const predicate = this.predicate();
    let found: InferResult;
    try {
      // @JAVA_REF org.modelingvalue.nelumbo.logic.Predicate#infer() no-arg:
      // Java does: Predicate p = setBinding(getBinding()); return p.resolve(context);
      const binding = predicate.getBinding();
      const bound = binding !== null ? predicate.setBinding(binding) as Predicate : predicate;
      const result = bound.resolve(knowledgeBase.createInferContext());
      const predicateResult = result.predicate(predicate);
      found = predicateResult as InferResult;
    } catch (e) {
      if (e instanceof Error) {
        handler.addException(ParseException.fromElements(e.message, predicate));
      }
      return;
    }
    this._inferResult = found;

    if (this.hasExpected()) {
      const trueBindings = this.expectedFacts();
      const truePredicates = trueBindings.map(b => predicate.setBinding(b)).toSet();
      const completeFacts = this.completeFacts();
      const falseBindings = this.expectedFalsehoods();
      const falsePredicates = falseBindings.map(b => predicate.setBinding(b)).toSet();
      const completeFalsehoods = this.completeFalsehoods();

      const expected = InferResult.ofWithPredicate(predicate, truePredicates, completeFacts, falsePredicates, completeFalsehoods, Set());

      if (!found.equals(expected) && found.toString() !== expected.toString()) {
        const astElements = this.astElements();
        handler.addException(
          ParseException.fromElements(
            'Expected result ' + expected + ', found ' + found,
            ...astElements.slice(2).toArray()
          )
        );
      }
    }
  }

  /**
   * Get the inference result.
   */
  inferResult(): InferResult | null {
    return this._inferResult;
  }
}
